### DISCUSSION TITLE
OPEN_CONTENTで新しいコンテンツを開いたときのTCP/IPによるソケット接続について

### DISCUSSION THREAD

**Discussion by kaz-wan** [mmdagent-ex/MMDAgent-EX#13] [Q&A]
引き続きいろいろ触らせていただいておりますが、バグか仕様かわからないものがありましたので、またご教授いただくことはできますでしょうか。

MMDAgent-EX（ver1.0.3）にTCP/IPで接続(MMDAget-EX側がサーバー)を行い、OPEN_CONTENTで新しいコンテンツを開くと、ソケットが一回閉じた後開くようなのですが、それは仕様でしょうか？

またその後、もう一度ソケット接続を行い、SPEAKコマンドで再生をしようとすると、最初に少し音が鳴った後フリーズしてしまいました。こちらはバグでしょうか？

環境はWIndows10環境です。よろしくお願いいたします。

**Comment by LeeAkinobu** [comment]
OPEN_CONTENT は全てのモジュールをシャットダウンしてから指定の.mdfで再起動するという動作をしています。このため、接続中のソケットは一度閉じてしまいます。.mdf にそのコンテンツが必要なシステムの機能やプラグインのON/OFFが書いてあるためこういう仕様になっています。

もし別の FST を起動するのが目的であれば、FST_LOAD を使うのも手かもしれません。

再起動後に SPEAK が動かないことがあるのはバグだと思います。少し調べてみます。

**Reply by kaz-wan** [reply]
ご回答ありがとうございます。

> OPEN_CONTENT は全てのモジュールをシャットダウンしてから指定の.mdfで再起動するという動作をしています。このため、接続中のソケットは一度閉じてしまいます。.mdf にそのコンテンツが必要なシステムの機能やプラグインのON/OFFが書いてあるためこういう仕様になっています。
もし別の FST を起動するのが目的であれば、FST_LOAD を使うのも手かもしれません。

なるほど、確かにそうですね。
FST_LOADで代用できるか検討してみます。

> 再起動後に SPEAK が動かないことがあるのはバグだと思います。少し調べてみます。

すいません、よろしくお願いいたします。

**Comment by LeeAkinobu** [comment]
再起動後に SPEAK_START が動かないバグを修正しました（初期化まわりの不具合でした）。

また、SPEAK で音声を再生するときに最後にノイズが入ることがあるバグも原因を見つけたので直しました（内部のフォーマット変換時の不具合でした）

動作確認いただけるとありがたいです。よろしくお願いします。

**Reply by kaz-wan** [reply]
回答遅くなりましたが、確認したところ問題なく動きました。ありがとうございます！

**Reply by kaz-wan** [reply]
ちなみに確認（要望）なのですが、現在SPEAKコマンドを使うと16kHzになるとのことですが、今後高音質化されていく予定とかはあったりしますでしょうか？

**Reply by LeeAkinobu** [reply]
ありがとうございます。内部のリップシンクモジュールの音声入力が16kHz固定で、再生と処理でバッファを共有しているので、16kHz に固定再生されてしまっています。全体を 48kHz 化するとか、バッファを分けてリップシンクだけ16kHzを使って再生は48kHzに並列化するなどすれば音質は改善できると思います。難易度がちょっと高いのですが、やはり再生音質は気になるのでTODOには入っている状態です。

**Reply by kaz-wan** [reply]
ご回答ありがとうございます。いずれ対応されるのを楽しみに待っています。
今回の質問事項は以上となります、ありがとうございました。

**Reply by LeeAkinobu** [reply]
対応した v1.0.4 をリリースしましたので、よろしければ試してみてください
